stan/math/rev/core/save_varis.hpp

```diff
@@ -0,0 +1,109 @@
+#ifndef STAN_MATH_REV_CORE_SAVE_VARIS_HPP
+#define STAN_MATH_REV_CORE_SAVE_VARIS_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Store a reference to `vari` used in `reduce_sum`.
+ * @tparam Pargs Types to be forwarded to recursive call of `save_varis`.
+ * @param[in, out] dest Pointer to set of varis.
+ * @param[in] x A `var` whose `vari` will be stores
+ * @param[in] args Further arguments forwarded to recursive call.
+ */
+template <typename... Pargs>
+inline vari** save_varis(vari** dest, const var& x, Pargs&&... args) {
+  *dest = x.vi_;
+  return save_varis(dest + 1, std::forward<Pargs>(args)...);
+}
```

Comment:
instead of an in/out and return should we just 
```suggestion
inline vari** save_varis(vari** dest, const var& x, Pargs&&... args) {
  *dest = x.vi_;
  save_varis(dest + 1, std::forward<Pargs>(args)...);
}
```

Reply:
No, I think the recursion is needed to carry on processing stuff... I am also not sure if your change is valid syntax given that there is no return, but the function declares that we return a `vari**`

Reply:
I don't think we use the return thing here. Same with accumulate_adjoints. But I intend to test it and I think it might be handy in the future for the reason @wds15 gives (already using these functions in the new ODE code).

Reply:
> No, I think the recursion is needed to carry on processing stuff... I am also not sure if your change is valid syntax given that there is no return, but the function declares that we return a vari**

The recursion will still happen, it's just that right now dest is both and in/out parameter and the return. I don't see a reason to have it as both

Reply:
I stuck with what we have. I like the return.

Reply:
> I like the return.

Why so? It feels odd to me that we'd have both a return and an in/out. We also call these like `save_vari(...)` without using the return

Reply:
Well we're only using dest as an input. I think it could be ```const vari** dest```.

Reply:
Oh we're writing to values point at by the pointer, but we're moving the pointer itself. It's like two different types of output.

Reply:
I guess ```vari** const``` is maybe the more appropriate type just judging from: https://stackoverflow.com/questions/1143262/what-is-the-difference-between-const-int-const-int-const-and-int-const ?

Reply:
The line here

```
*dest = x.vi_;
```

is modifying the input parameter, imo input parameters in general mean that the function does not modify the input argument (as if it was const). This does modify the input so I'd classify it as an in/out parameter. We also call this function as `save_vari(dest, blah...)`. Unless it's not doing anything then `dest` is an output parameter.

Reply:
Yeah I guess I just reverted to not thinking about consts and whatnot with the pointer types.

To achieve what the return is achieving though, I think we'd need the input to be vari***, cause what it's doing is returning a pointer to the vari pointer after the ones it wrote (so to send that info out we'd need to somehow communicate a vari** to the outside, which means a vari***).

Reply:
idt I'm following what the return is doing. The only place I see it used is when we loop over a container of containers ala

```cpp
template <typename VecContainer, require_std_vector_st<is_var, VecContainer>*,
          require_std_vector_vt<is_container, VecContainer>*, typename... Pargs>
inline vari** save_varis(vari** dest, VecContainer&& x, Pargs&&... args) {
  for (size_t i = 0; i < x.size(); ++i) {
    dest = save_varis(dest, x[i]);
  }
  return save_varis(dest, std::forward<Pargs>(args)...);
}
```

but that loop can just be replaced with

```cpp
  for (auto&& x_iter : x) {
    save_varis(dest, x_iter);
    dest += x_iter.size();
  }
```

I rewrote this in the branch I'm cleaning up rn to just use the above version with signature `void save_vari(...)`. is it okay if I push it up to the new branch / pr?

Reply:
Oh actually the same is true for accumulate adjoints no?

Reply:
This isn't the same thing:

```
  for (auto&& x_iter : x) {
    save_varis(dest, x_iter);
    dest += x_iter.size();
  }
```

What if x is a ```std::vector<std::vector<T>>```?

Reply:
I left the universal typedefs on ```std::vector<var>```s leave my damn returns alone lol. They're tested and everything.

Reply:
I think you could do:

```
for (auto&& x_iter : x) {
    save_varis(dest, x_iter);
    dest += count_vars(x_iter);
}
```

But I like the return thing better. It sorta gets the count itself which is nice.

```accumulate_adjoints``` is doing the same trick for the container version.

And yeah feel free to push straight to this branch.

Reply:
lol okay okay